src/utilities/createRouter.ts

```diff
@@ -45,18 +43,38 @@ export function createRouter<T extends Routes>(routes: T, options: RouterOptions
     Object.assign(route, newRoute)
   }
 
-  const push: RouterPush = async (url, options) => {
-    await navigation.update(url, options)
+  function pushUrl(url: string, options: RouterPushOptions = {}): Promise<void> {
+    return navigation.update(url, options)
+  }
+
+  function pushRoute({ name, params, replace }: { name: string, params?: Record<string, any> } & RouterPushOptions): Promise<void> {
+    const match = resolved.find((route) => flattenParentMatches(route) === name)
+
+    if (!match) {
+      throw `No route found with name "${String(name)}"`
+    }
+
+    const url = assembleUrl(match, params)
+
+    return navigation.update(url, { replace })
+  }
+
+  function push(urlOrRoute: string | { name: string, params?: Record<string, any> } & RouterPushOptions, possiblyOptions: RouterPushOptions = {}): Promise<void> {
+    if (typeof urlOrRoute === 'string') {
+      return pushUrl(urlOrRoute, possiblyOptions)
+    }
+
+    return pushRoute(urlOrRoute)
   }
 
-  const replace: RouterReplace = async (url) => {
-    await navigation.update(url, { replace: true })
+  async function replace(url: string, options: RouterReplaceOptions = {}): Promise<void> {
+    await navigation.update(url, { ...options, replace: true })
   }
 
   const router = {
-    routes: createRouteMethods<T>(resolved, push),
+    routes: createRouteMethods<T>(resolved, pushUrl),
     route: readonly(route),
-    push,
+    push: push as RouterPush<T>,
```

Comment:
This is pretty much the same as what we did for `createRouteMethods` where we just tell typescript what it is. We could probably abstract the implementation out into a `createRouterPush` utility that hides the type casting. 

Though with some fresh eyes I might be able to get rid of this.